wolfcrypt/src/ecc.c

```diff
@@ -1553,9805 +407,2360 @@ int wc_ecc_set_curve(ecc_key* key, int keysize, int curve_id)
+ */
+int wc_ecc_make_pub_ex(ecc_key* key, ecc_point* pubOut, WC_RNG* rng)
 {
-    int maxSigSz = 0;
-
-    /* calculate based on key bits */
-    /* maximum possible signature header size is 7 bytes plus 2 bytes padding */
-    maxSigSz = (sz * 2) + SIG_HEADER_SZ + ECC_MAX_PAD_SZ;
-
-    /* if total length is less than 128 + SEQ(1)+LEN(1) then subtract 1 */
-    if (maxSigSz < (128 + 2)) {
-        maxSigSz -= 1;
-    }
+    WOLFSSL_ENTER("wc_ecc_make_pub");
 
-    return maxSigSz;
+    return ecc_make_pub_ex(key, NULL, pubOut, rng);
 }
 
-/* maximum signature size based on actual key curve */
-int wc_ecc_sig_size(ecc_key* key)
+
+int wc_ecc_make_key_ex2(WC_RNG* rng, int keysize, ecc_key* key, int curve_id,
+                        int flags)
 {
-    int maxSigSz;
-    int orderBits, keySz;
 
-    if (key == NULL || key->dp == NULL)
-        return 0;
+    int err;
 
-    /* the signature r and s will always be less than order */
-    /* if the order MSB (top bit of byte) is set then ASN encoding needs
-        extra byte for r and s, so add 2 */
-    keySz = key->dp->size;
-    orderBits = wc_ecc_get_curve_order_bit_count(key->dp);
-    if (orderBits > keySz * 8) {
-        keySz = (orderBits + 7) / 8;
+    if (key == NULL || rng == NULL) {
+        return BAD_FUNC_ARG;
     }
-    /* maximum possible signature header size is 7 bytes */
-    maxSigSz = (keySz * 2) + SIG_HEADER_SZ;
-    if ((orderBits % 8) == 0) {
-        /* MSB can be set, so add 2 */
-        maxSigSz += ECC_MAX_PAD_SZ;
+
+    /* make sure required variables are reset */
+    wc_ecc_reset(key);
+
+    err = wc_ecc_set_curve(key, keysize, curve_id);
+    if (err != 0) {
+        return err;
+    }
+
+    key->flags = flags;
+
+#ifdef WOLFSSL_HAVE_SP_ECC
+#ifdef WOLFSSL_SP_384
+    if (key->idx != ECC_CUSTOM_IDX && ecc_sets[key->idx].id == ECC_SECP384R1) {
+        err = sp_ecc_make_key_384(rng, &key->k, &key->pubkey, key->heap);
+        if (err == MP_OKAY) {
+            key->type = ECC_PRIVATEKEY;
+        }
     }
-    /* if total length is less than 128 + SEQ(1)+LEN(1) then subtract 1 */
-    if (maxSigSz < (128 + 2)) {
-        maxSigSz -= 1;
+    else
+#endif
+#endif /* WOLFSSL_HAVE_SP_ECC */
+
+   { /* software key gen */
+        err = WC_KEY_SIZE_E;
     }
 
-    return maxSigSz;
+    return err;
 }
 
+WOLFSSL_ABI
+int wc_ecc_make_key_ex(WC_RNG* rng, int keysize, ecc_key* key, int curve_id)
+{
+    return wc_ecc_make_key_ex2(rng, keysize, key, curve_id, WC_ECC_FLAG_NONE);
+}
 
-#ifdef FP_ECC
-
-/* fixed point ECC cache */
-/* number of entries in the cache */
-#ifndef FP_ENTRIES
-    #define FP_ENTRIES 15
+#ifndef WOLFSSL_DO178_MAX_ECC_KEY
+    #define WOLFSSL_DO178_MAX_ECC_KEY 4
 #endif
+static byte ecc_key_buffer[WOLFSSL_DO178_MAX_ECC_KEY][sizeof(ecc_key)];
+static int ecc_key_in_use[WOLFSSL_DO178_MAX_ECC_KEY] = {0};
 
-/* number of bits in LUT */
-#ifndef FP_LUT
-    #define FP_LUT     8U
+WOLFSSL_ABI
+ecc_key* wc_ecc_key_new(void* heap)
```

Comment:
Does the customer really need these wc_ecc_key_new() and wc_ecc_key_free() API's?